zerver/lib/user_topics.py

```diff
@@ -95,39 +98,51 @@ def set_topic_mutes(
         recipient_id = stream.recipient_id
         assert recipient_id is not None
 
-        add_topic_mute(
+        set_user_topic_visibility_policy(
             user_profile=user_profile,
             stream_id=stream.id,
             recipient_id=recipient_id,
             topic_name=topic_name,
-            date_muted=date_muted,
+            visibility_policy=UserTopic.MUTED,
+            date_added=date_muted,
         )
 
 
-def add_topic_mute(
+@transaction.atomic
+def set_user_topic_visibility_policy(
     user_profile: UserProfile,
     stream_id: int,
-    recipient_id: int,
     topic_name: str,
-    date_muted: Optional[datetime.datetime] = None,
+    *,
+    visibility_policy: int,
+    recipient_id: int,
+    date_added: Optional[datetime.datetime] = None,
     ignore_duplicate: bool = False,
 ) -> None:
-    if date_muted is None:
-        date_muted = timezone_now()
-    UserTopic.objects.bulk_create(
-        [
-            UserTopic(
-                user_profile=user_profile,
-                stream_id=stream_id,
-                recipient_id=recipient_id,
-                topic_name=topic_name,
-                last_updated=date_muted,
-                visibility_policy=UserTopic.MUTED,
-            ),
-        ],
-        ignore_conflicts=ignore_duplicate,
+    if date_added is None:
+        date_added = timezone_now()
+
+    (row, created) = UserTopic.objects.get_or_create(
+        user_profile=user_profile,
+        stream_id=stream_id,
+        topic_name__iexact=topic_name,
+        recipient_id=recipient_id,
+        defaults={
+            "topic_name": topic_name,
+            "last_updated": date_added,
+            "visibility_policy": visibility_policy,
+        },
     )
 
+    if not created:
+        if row.visibility_policy == visibility_policy and not ignore_duplicate:
+            # Duplicate request case (topic-visibility_policy pair already exists)
+            raise JsonableError(_("Topic already muted"))
```

Comment:
While re-reviewing, I found that we can't hardcode the error message as "Topic already muted".
It can be "Topic already unmuted", "Topic already followed" etc.

Also, we didn't get this error means, we need to update our tests for unmute-topic to include this case.